plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchSearchEngine.ts

```diff
@@ -0,0 +1,274 @@
+        .requestBodySearch()
+        .query(esb.boolQuery().filter(filter).must([query]))
+        // TODO: Replace size limit with page cursor after pagination approach decided
+        // See: https://github.com/backstage/backstage/issues/6062
+        .size(100)
+        .toJSON(),
+      documentTypes: types,
+    };
+  }
+
+  setTranslator(translator: ElasticSearchQueryTranslator) {
+    this.translator = translator;
+  }
+
+  async index(type: string, documents: IndexableDocument[]): Promise<void> {
+    this.logger.info(
+      `Started indexing ${documents.length} documents for index ${type}`,
+    );
+    const startTimestamp = process.hrtime();
+    const alias = this.constructSearchAlias(type);
+    const index = this.constructIndexName(type, `${Date.now()}`);
+    try {
+      const aliases = await this.elasticSearchClient.cat.aliases({
+        format: 'json',
+        name: alias,
+      });
+      const removableIndices = aliases.body.map(
+        (r: Record<string, any>) => r.index,
+      );
+
+      await this.elasticSearchClient.indices.create({
+        index,
+      });
+      const result = await this.elasticSearchClient.helpers.bulk({
+        datasource: documents,
+        onDocument() {
+          return {
+            index: { _index: index },
+          };
+        },
+        refreshOnCompletion: index,
+      });
+
+      this.logger.info(
+        `Indexing completed for index ${type} in ${duration(startTimestamp)}`,
+        result,
+      );
+      await this.elasticSearchClient.indices.updateAliases({
+        body: {
+          actions: [
+            { remove: { index: this.constructIndexName(type, '*'), alias } },
+            { add: { index, alias } },
+          ],
+        },
+      });
+
+      this.logger.info('Removing stale search indices', removableIndices);
+      if (removableIndices.length) {
+        await this.elasticSearchClient.indices.delete({
+          index: removableIndices,
+        });
+      }
+    } catch (e) {
+      this.logger.error(`Failed to index documents for type ${type}`, e);
+      const response = await this.elasticSearchClient.indices.exists({
+        index,
+      });
+      const indexCreated = response.body;
+      if (indexCreated) {
+        this.logger.info(`Removing created index ${index}`);
+        await this.elasticSearchClient.indices.delete({
+          index,
+        });
+      }
+    }
+  }
+
+  async query(query: SearchQuery): Promise<SearchResultSet> {
+    const { elasticSearchQuery, documentTypes } = this.translator(query);
+    const queryIndices = documentTypes
+      ? documentTypes.map(it => this.constructSearchAlias(it))
+      : this.constructSearchAlias('*');
+    try {
+      const result = await this.elasticSearchClient.search({
+        index: queryIndices,
+        body: elasticSearchQuery,
+      });
+      return {
+        results: result.body.hits.hits.map((d: ElasticSearchResult) => ({
+          type: d._index.split('__')[0],
```

Comment:
Looks like this returns `[type]-index` instead of `[type]`, which results in all results falling back to the default result list item, rather than their specific typed versions.

Reply:
oh yeah, good catch. I'll add a bit more logic to this desctructuring.